packages/assets-controllers/src/balances.ts

```diff
@@ -0,0 +1,1103 @@
+      groupId: AccountGroupId,
+    ): AccountGroupBalance => ({
+      walletId,
+      groupId,
+      totalBalanceInUserCurrency: 0,
+      userCurrency: currencyRateState.currentCurrency,
+    });
+
+    flatAccountBalances.forEach((flatAccount) => {
+      const { walletId, groupId, balance } = flatAccount;
+      wallets[walletId as AccountWalletId] ??= defaultWalletBalance(
+        walletId as AccountWalletId,
+      );
+      wallets[walletId as AccountWalletId].groups[groupId as AccountGroupId] ??=
+        defaultGroupBalance(
+          walletId as AccountWalletId,
+          groupId as AccountGroupId,
+        );
+      wallets[walletId as AccountWalletId].groups[
+        groupId as AccountGroupId
+      ].totalBalanceInUserCurrency += balance;
+      wallets[walletId as AccountWalletId].totalBalanceInUserCurrency +=
+        balance;
+    });
+
+    // Ensure all groups (including empty ones) are represented
+    Object.entries(accountTreeState.accountTree.wallets ?? {}).forEach(
+      ([walletId, wallet]) => {
+        if (!wallet) {
+          return;
+        }
+        wallets[walletId as AccountWalletId] ??= defaultWalletBalance(
+          walletId as AccountWalletId,
+        );
+        Object.keys(wallet.groups || {}).forEach((groupId) => {
+          wallets[walletId as AccountWalletId].groups[
+            groupId as AccountGroupId
+          ] ??= defaultGroupBalance(
+            walletId as AccountWalletId,
+            groupId as AccountGroupId,
+          );
+        });
+      },
+    );
+
+    return wallets;
+  };
+
+  const flatAccounts = getFlatAccountBalances();
+  return {
+    wallets: getWalletBalances(flatAccounts),
+    totalBalanceInUserCurrency: getAggWalletBalance(flatAccounts),
+    userCurrency: currencyRateState.currentCurrency,
+  };
+}
+
+/**
+ * Calculate aggregated portfolio value change for a given period (1d, 7d, 30d).
+ * Logic mirrors extension/mobile historical aggregation:
+ * - For each asset with available percent change for the requested period, compute current value in user currency.
+ * - Reconstruct previous value by dividing current by (1 + percent/100).
+ * - Sum across all assets, then compute amount change and percent change.
+ *
+ * @param accountTreeState - AccountTreeController state.
+ * @param accountsState - AccountsController state.
+ * @param tokenBalancesState - TokenBalancesController state.
+ * @param tokenRatesState - TokenRatesController state.
+ * @param multichainRatesState - MultichainAssetsRatesController state.
+ * @param multichainBalancesState - MultichainBalancesController state.
+ * @param tokensState - TokensController state.
+ * @param currencyRateState - CurrencyRateController state.
+ * @param enabledNetworkMap - Map of enabled networks keyed by namespace.
+ * @param period - Period to compute change for ('1d' | '7d' | '30d').
+ * @returns Aggregated change details for the requested period.
+ */
+export function calculateBalanceChangeForAllWallets(
```

Comment:
Heya, did some cleanup on these functions in this diff
https://github.com/MetaMask/core/pull/6319

Feel free to look and pick through the commits (based off of this branch)